tests/unit_tests/test_mask.py

```diff
@@ -0,0 +1,374 @@
+        "data": result_3bands_3dates,
+        "shape": [1, 3, 3],
+        "stride": [9, 3, 1],
+        "offset": 0,
+    },
+}
+
+
+@pytest.mark.parametrize(
+    "example_input, additional_js_code_specific_to_case, expected_output",
+    [
+        (  # test different mask values: mask with number values
+            {
+                "data": data_3bands_3dates,
+                "mask": mask_3bands_3dates_num,
+                "scenes_data": scenes_3dates,
+                "scenes_mask": scenes_3dates,
+            },
+            None,
+            resultWithReplacement(result_3bands_3dates),
+        ),
+        (  # test different mask values: mask with boolean values
+            {
+                "data": data_3bands_3dates,
+                "mask": mask_3bands_3dates_bool,
+                "scenes_data": scenes_3dates,
+                "scenes_mask": scenes_3dates,
+            },
+            None,
+            resultWithReplacement(result_3bands_3dates),
+        ),
+        (  # test different replacement values: replacement = number
+            {
+                "data": data_3bands_3dates,
+                "mask": mask_3bands_3dates_num,
+                "scenes_data": scenes_3dates,
+                "scenes_mask": scenes_3dates,
+                "replacement": replacement_val_num,
+            },
+            None,
+            resultWithReplacement(result_3bands_3dates, replacement_val_num),
+        ),
+        (  # test different replacement values: replacement = boolean
+            {
+                "data": data_3bands_3dates,
+                "mask": mask_3bands_3dates_num,
+                "scenes_data": scenes_3dates,
+                "scenes_mask": scenes_3dates,
+                "replacement": replacement_val_bool,
+            },
+            None,
+            resultWithReplacement(result_3bands_3dates, replacement_val_bool),
+        ),
+        (  # test different replacement values: replacement = string
+            {
+                "data": data_3bands_3dates,
+                "mask": mask_3bands_3dates_num,
+                "scenes_data": scenes_3dates,
+                "scenes_mask": scenes_3dates,
+                "replacement": replacement_val_str,
+            },
+            None,
+            resultWithReplacement(result_3bands_3dates, replacement_val_str),
+        ),
+        (  # test mask parameter missing dimensions: mask parameter with no temporal dimension
+            {
+                "data": data_3bands_3dates,
+                "mask": [mask_3bands_3dates_num[0]],
+                "scenes_data": scenes_3dates,
+                "scenes_mask": None,
+            },
+            "maskCube.removeDimension('temporal_name');",
+            resultWithReplacement(result_mask_no_temporal),
+        ),
+        # (  # test mask parameter missing dimensions: mask parameter with no bands dimension
+        #     # doesn't work properly: same result as for "mask parameter with no temporal dimension"
```

Comment:
We should "help" the broadcasting method by restricting the options how the smaller array can be broadcasted in the bigger one.

`[0,11,0]` could be validly broadcasted into an array of shape `(3,3)` either row-wise
```js
[
  [0, 11, 0],
  [0, 11, 0],
  [0, 11, 0]
]
```
or column-wise
```js
[
  [ 0,  0,  0],
  [11, 11, 11],
  [ 0,  0,  0]
]
```
and the method can't know we want the second one in this case.

We can help it by inserting the missing dimensions into the `mask` as a label-less dimensions of length 1. So in our case  `[0,11,0]` would become `[[0],[11],[0]]` or `ndarray([0,11,0], [3,1])`. Now we only one option for broadcasting.

Reply:
Thanks, I added inserting of the missing dimensions in the `mask` argument as label-less dimensions of length 1.

I also added ordering the dimensions in the `mask` argument so that they are in the same order as in the `data` argument. This then guarantees that positions of the elements in the flattened arrays of both arguments are synced in terms of dimensions. (the solution for the [comment below](https://github.com/openEOPlatform/openeo-pg-evalscript-converter/pull/63#discussion_r869269031))

Reply:
Wonderful, can we please add a test for this?

Reply:
I added the test and while making it, I realized that there was a bug in `flattenToNativeArray()` function.

I removed the check for the subarray (code below) because with this check, it didn't use the `shape` and `stride` which are important for the transposed ndarrays - when the ndarray is transposed (`ndarray.transpose()`), only the `shape` and `stride` are changed, the `data` of the ndarray stays the same - and thus returned the wrong result.

```
if (isNotSubarray(ndarray, shape)) {
  return ndarray.data
}
```

This might affect the behaviour of the other processes that use `flattenToNativeArray()` / `DataCube.flattenToArray()`, but the tests are all successful, so it doesn't seem this result in a wrong things to be returned for other processes.

Reply:
Well, we never had a situation like that so `flattenToNativeArray` wasn't designed to handle it. However, simply returning the array is surely much faster that iterating over it. So I'd prefer if we added a flag (`transposed=true` or whatever), so that the simple approach would still be used, as the majority of the cases should use it. 

Of course we should avoid ever returning a transposed array from a process, but that's already the case now as other processes do not expect it.

Of course we can add `.slice()` for consistency.

-----------------------------------

Other than that, are you sure the tests work correctly? I find them a bit hard to read (I prefer the approach of e.g. `test_merge_cube`, i.e. passing data, shape and dimensions and setting them directly, but it's not a big deal) so I'm not sure if I'm misunderstanding them or not. 

Given it's being transposed and the dimensions switched after initialisation, I assume `maskCube` data should be the same as in the previous test (instead of that `mask_3bands_3dates_num[0],`). Because it's essentially transposed twice then.

Reply:
> Ok I think they seem ok, It's setup as every second band to be masked, then transposed to every second temporal element.

Yes, it was like that.
I agree that it was hard to read. I changed the approach to the same as in `test_merge_cube.py`.

Reply:
Looks good.